File: Java/Object Validator/lib/src/main/java/com/quicksilver/objectvalidator/runtime/Or.java
package com.quicksilver.objectvalidator.runtime;

import java.util.HashSet;

public class Or extends Condition {
    protected final Condition[] conditions;

    public Or(boolean reversed, String fieldExpression, Condition... conditions) {
        super(reversed, fieldExpression);
        this.conditions = conditions;
    }

    @Override
    protected boolean isFulfilledBy(Object field, String fullFieldExpression, HashSet<String> passedFields, HashSet<String> failedFields) throws ReflectiveOperationException {
        HashSet<String> newFailedFields = new HashSet<>();
        for (Condition condition : conditions)
            if (condition.check(field, fullFieldExpression, passedFields, newFailedFields))
                return true;
        failedFields.addAll(newFailedFields);
        return false;
    }
}

File: C Sharp/Object Validator/Runtime/Or.cs
namespace Quicksilver.ObjectValidator.Runtime;
class Or(bool reversed, string? fieldExpression, Condition[] conditions) : Condition(reversed, fieldExpression)
{
    protected readonly Condition[] conditions = conditions;

    protected override bool IsFulfilledBy(object? value, string? fullFieldExpression, HashSet<string> passedFields, HashSet<string> failedFields)
    {
        HashSet<string> newFailedFields = [];
        foreach (Condition condition in conditions)
            if (condition.Check(value, fullFieldExpression, passedFields, newFailedFields))
                return true;
        failedFields.UnionWith(newFailedFields);
        return false;
    }
}